lib/elixir/lib/application.ex

```diff
@@ -412,12 +439,154 @@ defmodule Application do
     :application.get_all_env(app)
   end
 
+  @doc """
+  Gets all compile env listeners.
+
+  A compile env listener is a process that receives a message in the shape of
+  `{:compile_env, compile_env_notification}`, with `compile_env_notification`
+  given by the type `t:compile_env_notification/0`, every time `compile_env/3`
+  or `compile_env!/2` is called.
+
+  Listeners can be added with `put_compile_env_listener/1` and removed
+  with `delete_compile_env_listener/1`.
+  """
+  @doc since: "1.10.0"
+  @spec get_compile_env_listeners() :: [pid]
+  def get_compile_env_listeners() do
+    :elixir_config.get(:compile_env_listeners)
+  end
+
+  @doc """
+  Puts a `pid` as a compile env listener.
+
+  See `get_compile_env_listeners/0` for more information.
+  """
+  @doc since: "1.10.0"
+  @spec put_compile_env_listener(pid) :: :ok
+  def put_compile_env_listener(pid) when is_pid(pid) do
+    :elixir_config.update(:compile_env_listeners, &[pid | &1])
+    :ok
+  end
+
+  @doc """
+  Deletes a `pid` from compile env listeners.
+
+  See `get_compile_env_listeners/0` for more information.
+  """
+  @doc since: "1.10.0"
+  @spec delete_compile_env_listener(pid) :: :ok
+  def delete_compile_env_listener(pid) when is_pid(pid) do
+    :elixir_config.update(:compile_env_listeners, &List.delete(&1, pid))
+    :ok
+  end
+
+  @doc """
+  Reads the application environment at compilation time.
+
+  Similar to `get_env/3`, except it must be used to read values
+  at compile time. This allows Elixir to track when configuration
+  values between compile time and runtime differ.
```

Comment:
Initially misread this sentence and was about to suggest a fix, then noticed it's correct but I think it reads clearer this way, feel free to ignore (only submitting this comment since you can just click "commit" if you like it 😄).

```suggestion
  values change between compile time and runtime.
```